Improve Reader empty state based on following status

pfefferle opened the pull request:
Fixes #

## Proposed changes:
* Add contextual empty state messages in the Reader based on whether the user is following anyone
* When not following anyone: show engaging message with link to the following page (URL differs for blog vs user actors)
* When following but no posts: show message about checking back for updates
* Add `followed_by` REST API parameter for `ap_actor` post type to query actors a user follows
* Add `useFollowing` hook to fetch following count via REST API
* Add unit tests for the new hook

<img width="1220" height="887" alt="Screenshot 2025-12-15 at 17 54 26" src="https://github.com/user-attachments/assets/028e2608-d086-4daf-b6e5-c5c484c0d144" />


### Other information:

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions:

* Go to the ActivityPub Reader page (`/wp-admin/index.php?page=activitypub-social-web`)
* With no followers, you should see "Your feed is waiting to come alive. Start following people on the Fediverse" with a link
* The link should go to:
  - `options-general.php?page=activitypub&tab=following` for the blog actor (ID 0)
  - `users.php?page=activitypub-following-list` for user actors
* Follow someone, then the message should change to "Nothing new from the people you follow. Check back soon for fresh updates."

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [x] Patch

#### Type

-   [x] Changed - for changes in existing functionality

#### Message

Improve Reader empty state to show contextual messages based on following status.

</details>

pfefferle committed:
Improve Reader empty state based on following status

- Add `followed_by` REST API parameter for querying actors a user follows
- Create `useFollowing` hook to fetch following count
- Update feed empty state to show different messages:
  - If not following anyone: link to the following page
  - If following but no posts: show existing message

matticbot committed:
Add changelog

pfefferle committed:
Merge branch 'add/reader' into update/reader-empty-state

pfefferle committed:
Update asset version in index.asset.php

pfefferle committed:
Merge branch 'add/reader' into update/reader-empty-state

pfefferle committed:
Merge branch 'add/reader' into update/reader-empty-state

pfefferle committed:
Refactor feed-content.js for improved modularity

Reorganized and renamed several variables and functions for clarity and modularity. Updated inspector and stage exports, restructured field and filter definitions, and improved the feed and inspector logic for better maintainability.

obenland commented on src/app/routes/feed/stage.tsx:
```diff
@@ -186,6 +193,32 @@ export default function FeedStage(): ReactNode {
+
+		// Default: following people but no posts yet.
+		return __( 'Nothing new from the people you follow. Check back soon for fresh updates.', 'activitypub' );
+	};
```
Let's extract renderEmptyStateContent() into its own EmptyState component. All of the things it depends on can be pulled through hooks.

obenland commented on src/app/routes/feed/stage.tsx:
```diff
@@ -186,6 +193,32 @@ export default function FeedStage(): ReactNode {
+		}
+
+		// If not following anyone, show link to following page.
+		if ( followingResolved && ! followingCount ) {
```
Hm, this adds quite a bit of complexity where it might not be strictly needed and we might get a better user experience without it.

Even if someone already follows a (single) user, there's still a good change they don't post frequently enough to fill a timeline, especially when starting out. Could we generalize the prompt to follow _more_ people instead?

pfefferle replied:
You mean one single text, that does not care why the list is empty and always show the follow (more) people link!?

obenland replied:
Unless there are filters active, yes. Maybe something like [my suggestion below](https://github.com/Automattic/wordpress-activitypub/pull/2634#discussion_r2620487259)?

obenland commented on src/app/routes/feed/stage.tsx:
```diff
@@ -186,6 +193,32 @@ export default function FeedStage(): ReactNode {
+			return (
+				<>
+					{ __( 'Your feed is waiting to come alive.', 'activitypub' ) }{ ' ' }
+					<a href={ followingUrl }>{ __( 'Start following people on the Fediverse', 'activitypub' ) }</a>
```
```suggestion
					{ createInterpolateElement(
						__(
							'Your feed is waiting to come alive. <a>Follow more people on the Fediverse</a> to see their posts here.',
							'activitypub'
						),
						{
							a: <a href={ followingUrl } />,
						}
					) }
```
Using `createInterpolateElement` would make this easier to translate for RTL languages.

obenland commented on src/app/routes/feed/stage.tsx:
```diff
@@ -301,16 +334,7 @@ export default function FeedStage(): ReactNode {
-						  ) }
-				</p>
-			}
+			empty={ <p>{ renderEmptyStateContent() }</p> }
```
```suggestion
			empty={ <EmptyState /> }
```
We should be able to pass a zero-prop component here.

pfefferle committed:
Simplify empty state to separate EmptyState component

Address PR feedback:
- Extract empty state logic into EmptyState component
- Remove following count check - always show "follow more" prompt
- Use createInterpolateElement for better RTL translation support
- Remove unused useFollowing hook and PHP followed_by parameter
- Simplify to single message for empty feed (except when filters active)

obenland commented on includes/class-post-types.php:
```diff
@@ -627,32 +627,6 @@ public static function register_ap_actor_rest_field() {
-			);
-		}
-
-		return $args;
```
This seems unrelated to the PR? The callback is needed for the Followers list that's currently disabled.

pfefferle replied:
PR incoming

pfefferle replied:
This should no longer be needed!

obenland replied:
Which PR was that?

My understanding is that the callback is needed for the Followers list that's currently disabled.

pfefferle replied:
Ah, I see! Sorry!

obenland reviewed (commented):
Nice, much cleaner!

pfefferle committed:
Merge branch 'add/reader' into update/reader-empty-state

pfefferle committed:
Fix useEntityRecords call to enable fetching by userId

Updated the useEntityRecords call in the feed-content.js build to pass the 'enabled' option when a userId is present. This ensures that entity records are only fetched when the userId is defined, preventing unnecessary or incorrect data fetching.

pfefferle committed:
Merge branch 'add/reader' into update/reader-empty-state

pfefferle committed:
Update sidebar icon references and translations

Replaces some sidebar icon references and translation function calls in build/app/index.js to use updated variable names and translation utilities. Also updates the asset version in index.asset.php.

pfefferle committed:
Merge branch 'add/reader' into update/reader-empty-state

pfefferle committed:
Fix tag filter aria-label in sidebar menu

Corrects the aria-label for tag filter menu items in the sidebar from 'Filter by type' to 'Filter by tag', improving accessibility and accuracy.

pfefferle committed:
Add follower_of filter to REST actor query

Introduces a custom query parameter 'follower_of' to the REST API actor query, allowing filtering of actors by follower relationships using a meta query. This enhances API flexibility for client applications needing to query actors based on their followers.

pfefferle commented:
@obenland done!
